data/covering_grammar/lib/error_analysis.py

```diff
@@ -19,51 +19,58 @@
 
 import argparse
 
-import prettytable  # type: ignore
+import prettytable
 import pynini
+from pynini.lib import rewrite
 
 
 def main(args: argparse.Namespace) -> None:
-    cg_fst = pynini.string_file(args.cg_path).closure().optimize()
-    rulematch_predmatch = 0
-    rulematch_pred_notmatch = 0
-    not_rulematch_predmatch = 0
-    not_rulematch_pred_notmatch = 0
-    total_records = 0
-    with open(args.test_path, "r") as source:
-        for line in source:
-            total_records += 1
-            parts = line.split("\t")
-            lg = parts[0].rstrip()
-            act = parts[1].replace(" ", "").replace(".", "").rstrip()
-            pred_pron = parts[2].replace(" ", "").replace(".", "").rstrip()
-            # TODO: use Pynini's rewrite module here.
-            lattice = lg @ cg_fst @ pred_pron
-            lattice.project("output")
-            if lattice.start() == pynini.NO_STATE_ID:
-                if act == pred_pron:
-                    not_rulematch_predmatch += 1
+    with pynini.default_token_type("utf8"):
+        cg_fst = pynini.string_file(args.cg_path).closure().optimize()
+        rulematch_predmatch = 0
+        rulematch_pred_notmatch = 0
+        not_rulematch_predmatch = 0
+        not_rulematch_pred_notmatch = 0
+        total_records = 0
+        with open(args.test_path, "r") as source:
+            for line in source:
+                total_records += 1
+                (ortho_str, hypo_p_str, gold_p_str) = line.rstrip().split(
+                    "\t", 2
+                )
+                ortho = pynini.accep(ortho_str)
```

Comment:
I don't think you need to explicitly convert any of these to an FST since the code is running within the UTF-8 context manager. `rewrite.matches` ought to do it for you. (If I'm wrong, though I'd like to know, maybe that's something I could fix in the future. Then you can do something like:

```python
ortho, hypo_p, gold_p = line.rstrip().split("\t", 2)
if rewrite.matches(ortho, hypo_p, cg_fst):
   ...
```

Apologies in advance if I'm wrong.

I would also recommend comparing the strings (e.g., on lines 45 and 50) rather than comparing the FSTs. I'm sure comparing two FSTs is much slower than comparing two strings.

Reply:
Fixed.